site/_posts/2023-04-13-securing-react-native-with-oauth-0-71.md

```diff
@@ -0,0 +1,501 @@
+
+The `redirectUrl` is the URL that we set up in the FusionAuth application, with the scheme we used in configuring iOS and Android. The value `oauthredirect` is a callback path defined by the React Native app auth library. Make sure you update the `issuer` and `clientId` keys in this object with your configuration values.
+
+We can also add any additional parameters (none, in this case). If you need custom scopes, this is the place to add them as well. We're requesting the `offline_access` scope so that the OAuth server will return a `refresh_token`. Such a token can be used to request additional access tokens should our current one expire.
+
+Next, create a default auth state object in the file. This will be modified as our user first views the React Native app, then authenticates. This contains information like the token values and expiration dates.
+
+```typescript
+//...
+const defaultAuthState = {
+  hasLoggedInOnce: false,
+  accessToken: '',
+  accessTokenExpirationDate: '',
+  refreshToken: ''
+};
+
+const [authState, setAuthState] = useState(defaultAuthState);
+//...
+```
+
+Now, we are ready to configure the code which receives the token. 
+
+## Configuring React Native OAuth authorization
+
+Let's create the function to get the token; this will use the previously created `configs` object. It will also use the `authorize` function from the `react-native-app-auth` package. It will do all the heavy lifting and connect with the OAuth server. The implementation of the function is below:
+
+```typescript
+//...
+const handleAuthorize = async () => {
+  try {
+    const newAuthState = await authorize(configs.fusionauth);
+
+    setAuthState({
+      hasLoggedInOnce: true,
+      ...newAuthState,
+    });
+  } catch (error) {
+    Alert.alert(
+      'Failed to log in',
+      (error as Record<string, never>)?.message,
+    );
+  }
+};
+//...
+```
+
+`newAuthState` is returned from the `authorize` function, as we can set our auth state to that returned value. Now we have the code to interface with FusionAuth, so we'll want to give the user a way to invoke the code.
+
+## Building the user interface
+
+So, we need to create a user interface (UI). We'll create a simple UI to begin authentication. After the user has logged in, we'll display the access token. The access token is what FusionAuth provides once a user has successfully signed in. 
+
+Of course, you typically don't want to simply display or store the access token. You want it because it allows your application to make other API calls, often to gather more information to display to the user. Later in this tutorial, we'll use an access token to retrieve user information from an OpenID Connect endpoint, and display that in our application.
```

Comment:
```suggestion
Of course, you typically don't want to simply display or store the access token. You want it because it allows your application to make other API calls, often to gather more information to display to the user. Later in this tutorial, we'll use an access token to retrieve user information from an OpenID Connect endpoint, and display that in our application. In a real application, you'd present the access token to other APIs to gain, well, access to data and functionality.
```